# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxPlatformFontList.cpp
# Commit: 40fcc0cdc593
# Full Hash: 40fcc0cdc593fa337273631de435a2901ec881f6
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2024-02-07 04:17:40
# Regressor Bug: 1878286
# File Overlap Count: 2
# Description:
#   Bug 1878286 - Eliminate usage of fontlist::Pointer in IPDL. r=gfx-reviewers,lsalzman
#   
#   No change in behavior.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D200564
# ==============================================================================

diff -r c48ca61bc925 -r 40fcc0cdc593 gfx/thebes/gfxPlatformFontList.cpp
--- a/gfx/thebes/gfxPlatformFontList.cpp	Tue Feb 06 10:15:29 2024 +0000
+++ b/gfx/thebes/gfxPlatformFontList.cpp	Tue Feb 06 10:41:50 2024 +0000
@@ -3051,7 +3051,8 @@
 }
 
 void gfxPlatformFontList::SetCharacterMap(uint32_t aGeneration,
-                                          const fontlist::Pointer& aFacePtr,
+                                          uint32_t aFamilyIndex, bool aAlias,
+                                          uint32_t aFaceIndex,
                                           const gfxSparseBitSet& aMap) {
   MOZ_ASSERT(XRE_IsParentProcess());
   auto list = SharedFontList();
@@ -3065,14 +3066,35 @@
   if (AppShutdown::IsInOrBeyond(ShutdownPhase::AppShutdownConfirmed)) {
     return;
   }
-  auto* face = aFacePtr.ToPtr<fontlist::Face>(list);
-  if (face) {
+
+  const fontlist::Family* family;
+  if (aAlias) {
+    if (aFamilyIndex >= list->NumAliases()) {
+      MOZ_ASSERT(false, "AliasFamily index out of range");
+      return;
+    }
+    family = list->AliasFamilies() + aFamilyIndex;
+  } else {
+    if (aFamilyIndex >= list->NumFamilies()) {
+      MOZ_ASSERT(false, "Family index out of range");
+      return;
+    }
+    family = list->Families() + aFamilyIndex;
+  }
+
+  if (aFaceIndex >= family->NumFaces()) {
+    MOZ_ASSERT(false, "Face index out of range");
+    return;
+  }
+
+  if (auto* face =
+          family->Faces(list)[aFaceIndex].ToPtr<fontlist::Face>(list)) {
     face->mCharacterMap = GetShmemCharMap(&aMap);
   }
 }
 
-void gfxPlatformFontList::SetupFamilyCharMap(
-    uint32_t aGeneration, const fontlist::Pointer& aFamilyPtr) {
+void gfxPlatformFontList::SetupFamilyCharMap(uint32_t aGeneration,
+                                             uint32_t aIndex, bool aAlias) {
   MOZ_ASSERT(XRE_IsParentProcess());
   auto list = SharedFontList();
   MOZ_ASSERT(list);
@@ -3086,46 +3108,20 @@
     return;
   }
 
-  // aFamilyPtr was passed from a content process which may not be trusted,
-  // so we cannot assume it is valid or safe to use. If the Pointer value is
-  // bad, we must not crash or do anything bad, just bail out.
-  // (In general, if the child process was trying to use an invalid pointer it
-  // should have hit the MOZ_DIAGNOSTIC_ASSERT in FontList::ToSharedPointer
-  // rather than passing a null or bad pointer to the parent.)
-
-  auto* family = aFamilyPtr.ToPtr<fontlist::Family>(list);
-  if (!family) {
-    // Unable to resolve to a native pointer (or it was null).
-    NS_WARNING("unexpected null Family pointer");
+  if (aAlias) {
+    if (aIndex >= list->NumAliases()) {
+      MOZ_ASSERT(false, "AliasFamily index out of range");
+      return;
+    }
+    list->AliasFamilies()[aIndex].SetupFamilyCharMap(list);
     return;
   }
 
-  // Validate the pointer before trying to use it: check that it points to a
-  // correctly-aligned offset within the Families() or AliasFamilies() array.
-  // We just assert (in debug builds only) on failure, and return safely.
-  // A misaligned pointer here would indicate a buggy (or compromised) child
-  // process, but crashing the parent would be unnecessary and does not yield
-  // any useful insight.
-  if (family >= list->Families() &&
-      family < list->Families() + list->NumFamilies()) {
-    size_t offset = (char*)family - (char*)list->Families();
-    if (offset % sizeof(fontlist::Family) != 0) {
-      MOZ_ASSERT(false, "misaligned Family pointer");
-      return;
-    }
-  } else if (family >= list->AliasFamilies() &&
-             family < list->AliasFamilies() + list->NumAliases()) {
-    size_t offset = (char*)family - (char*)list->AliasFamilies();
-    if (offset % sizeof(fontlist::Family) != 0) {
-      MOZ_ASSERT(false, "misaligned Family pointer");
-      return;
-    }
-  } else {
-    MOZ_ASSERT(false, "not a valid Family or AliasFamily pointer");
+  if (aIndex >= list->NumFamilies()) {
+    MOZ_ASSERT(false, "Family index out of range");
     return;
   }
-
-  family->SetupFamilyCharMap(list);
+  list->Families()[aIndex].SetupFamilyCharMap(list);
 }
 
 bool gfxPlatformFontList::InitOtherFamilyNames(uint32_t aGeneration,